File: serena_obrien/homework1/q5_ShortestSubstring.py
# Time complexity: O(n)
# Space complexity: O(m), where m is the number of unique characters in requiredChars

# Technique: Variable-size sliding window

def ShortestSubstring(s, requiredChars):
    if not s or not requiredChars:
        return 0

    minLength = float("inf")
    required = {}
    for char in requiredChars:
        required[char] = required.get(char, 0) + 1
    currWindowCounts = {char: 0 for char in required.keys()}
    progress = 0
    complete = len(required) # number of unique characters in requiredChars

    left = 0
    for right in range(0, len(s)):
        currChar = s[right]

        if currChar in required:
            currWindowCounts[currChar] += 1
            if currWindowCounts[currChar] == required[currChar]:
                progress += 1

        while progress == complete:
            minLength = min(minLength, right - left + 1)

            leftChar = s[left]
            if leftChar in required:
                currWindowCounts[leftChar] -= 1
                if currWindowCounts[leftChar] < required[leftChar]:
                    progress -= 1
        
            left += 1

    if minLength == float("inf"):
        return None
    return minLength

if __name__ == '__main__':
    # inputStr = "abracadabra"
    # requiredChars = "abc"
    inputStr = "zxycbaabcdwxyzzxwdcbxyzabccbazyx"
    requiredChars = "zzyzx"
    # inputStr = "do"
    # requiredChars = "god"
    # inputStr = ""
    # requiredChars = ""
    print("Input Strings:", inputStr, ',', requiredChars)
    print("Output:", ShortestSubstring(inputStr, requiredChars))

# ~ time spent: 35 minutes

File: homework1/nitaimahat/q5_ShortestSubstring.java
package homework1.nitaimahat;
import java.util.*;
public class q5_ShortestSubstring {
    // thought process: since i need to the shortest substring, im thinking of using array with
    // the leght of ascii letetrs which is 128 and thne i will increment the target string by 1 
    // i will have two pointers first second pointing at 0 index of the comparing string and then 
    // have a variable counter taht keeps tracks of number of character in target string in the comparing string
    // if the counter equals the size of the target string i will check for minwindow size}
    // and continue comapring and retun the minwindow size at the end
    // the time complexity will be o(n+m) since im using a slidding window pattern and each pointer moves at most n times 
    // and space will be o(n+m) since i will be creating character array from the input strings.
    public static int shortSubstring(String s1, String s2){
        int[] arr = new int[128];
        char[] firstArr = s1.toCharArray();
        char[] secondArr = s2.toCharArray();

        for(char curr : secondArr){
            arr[curr]++;
        }
        int first = 0;
        int second = 0;
        int counter = 0;
        int minLength = Integer.MAX_VALUE;

        while(second < firstArr.length){
            char currChar = firstArr[second];
            arr[currChar]--;
            if(arr[currChar] >=0){
                counter++;
            }
            //shringk window

            while(counter == secondArr.length){
                int currWindow = second - first +1;
                if(currWindow < minLength){
                    minLength = currWindow;
                }
                char firstCount = firstArr[first];
                arr[firstCount]++;
                if(arr[firstCount]>0){
                    counter--;
                }
                first++;
            }
            second++;
        }
        return minLength;

    }
    //time spent 32 min

    public static void main(String[] args){
        String s1 = "abracadabra";
        String s2 = "abc";
        System.out.println(shortSubstring(s1, s2));

         String s3 = "zxycbaabcdwxyzzxwdcbxyzabccbazyx";
        String s4 = "zzyzx";
        System.out.println(shortSubstring(s3, s4));

         String s5 = "dog";
        String s6 = "god";
        System.out.println(shortSubstring(s5, s6));


    }

}
